api/proto/objects/revision.proto

```diff
@@ -0,0 +1,70 @@
+syntax = "proto3";
+
+package jj_api.objects;
+
+// A RevisionMask controls how much of a revision is returned.
+// For example, `jj log` will likely not care about files, while `jj status`
+// will want to know what files changed in a revision, and `jj diff` will want
+// to know both the content of the file and the content of the file in the
+// parent revision.
+message RevisionMask {
+  // Which files will be returned.
+  enum FilePathMask {
+    NONE = 0;
+    MODIFIED_FILES = 1;
+    ALL_FILES = 3;
+    // Include all files that were matched by the RevisionMask containing this
+    // RevisionMask. For example, if we had:
+    // RevisionMask(
+    //   files_to_include = MODIFIED_FILES,
+    //   files = FileMask(content = True),
+    //   parents = RevisionMask(
+    //     files_to_include = PARENT_FILES
+    //     files = FileMask(content = True),
+    //   ),
+    // )
+    // And we returned the revision @, which modfied the file foo, then both
+    // r.files["foo"].content and r.parent.files["foo"].content would be filled.
+    PARENT_FILES = 4;
+  };
+  FilePathMask files_to_include = 1;
+  repeated string additional_files = 2;
+  
+  bool rendered = 3;
+
+  bool file_content = 4;
+  bool file_hash = 5;
+  bool file_metadata = 6;
+
+  // How much of the parent revision to fill in.
+  RevisionMask parents = 7;
+}
+
+message File {
+  // If you query for MODIFIED_FILES, you may get a file that's been deleted.
+  bool exists = 1;
+  string hash = 2;
+  bytes content = 3;
+  // TODO: file metadata (eg. permissions)?
+}
+
+
+// By default, 
+message Revision {
+  string change_id = 1;
+  string commit_id = 2;
+  string description = 3;
+  bool empty = 4;
+  bool conflicts = 5;
+  bool mutable = 6;
+  
+  // The rendered template.
+  // Open question: Is a template a first-class citizen of jj, or is it specific
+  // to jj-cli?
```

Comment:
Interesting question! My gut feeling is that templating should not be in the API.

However, if we return all the data any caller might need, it might get unnecessarily slow. For example, it's currently expensive to check if merge commits are empty (because we check that by merging the parents' trees). So we may want to be able to indicate which data we want back. There may be some value in being able to make that conditional like you can in a template (e.g. only care whether a commit is empty if it's mutable). That suggests something like templates, but I'm not sure it will be needed. Maybe callers that care enough can send multiple requests instead (e.g. send an additional request for each mutable commit). Even if we provide some support for conditionals like that, I don't think we should return a *rendered* string. The the caller might have to parse it. It seems better to return it in some structured form.

Reply:
> However, if we return all the data any caller might need, it might get unnecessarily slow.

I'm open to adding fields to the `RevisionMask` proto.

> There may be some value in being able to make that conditional like you can in a template (e.g. only care whether a commit is empty if it's mutable).

I'm not a huge fan of this, this seems unnecessarily complex, and I can't see a use case for it. I think that, for example, if the user wanted to retrieve the diff, and thus only wanted to know if the current commit was empty, but needed the parent commit, then they can simply do a `GetRevision` on the current revision, and add a mask for the parents so that the parents and the revision you actually care about return different sets of data.

Reply:
I was thinking of the current CLI as an example user of the API, but I think our conclusion from the design doc was that the current CLI needs more control than we expect to provide via the RPC API.

Reply:
Ah, sounds like we both understood the conclusion differently here. I understood our conclusion to be that the current CLI currently needs more control than we can expect to provide via the RPC API, but the conclusion I drew was that we would have to do significant refactorings to the current CLI in order to change that. I think, however, that your conclusion is probably better for the short-term, and for now we shouldn't try to migrate jj-cli over.

That being said, I think we should design this with the assumption that if jj-cli needs more control than we expect to provide, then other users may need that as well. Thus, I think we should design this API with the intent that if jj-cli requires more than this API can provide right now, we should attempt to design it so that we can at least add those features to the API in the future so we could migrate.

Reply:
Parallelize doesn't seem too difficult, TBH (maybe I'm missing something though). To me, the hard part wasn't the actual performing of any of these operations, it's the fact that jj-cli reports back secondary information (eg. the squash command reports "rebased x child commits"):

```python
def parallelize(revsets: list[str]):
  transaction = client.BeginTransaction()
  revisions = client.ListRevisions(revset=Revset(union=[Revset(revset=x) for x in revsets]
  
  selected = {r.change_id for r in revisions}
  parents = {parent.change_id for r in revisions for parent in r.parents} - selected
  # I think adding a "children" field to a revision seems reasonable (assuming we mask it out by default)
  children = {child.change_id for r in revisions for child in r.children} - selected
  
  new_parents = {}
  # Rebase the selected nodes onto all nodes that were parents of any of them
  for r in selected:
      new_parents[r] = parents
  # Rebase the child nodes onto all nodes in this revset.
  for r in children:
      new_parents[r] = selected
  
  client.Rebase(new_parents = new_parents)
  client.EndTransaction()
```

Reply:
The parts I were thinking of as hard were the transaction management and the callback in `transform_descendants()`. I think we previously said the API would be stateless. I see that that has now changed with the transaction methods. Callbacks might be implementable using streaming RPCs. Maybe we'll implement some such state management using `RefCell` and it won't be much of a problem. I'm sure it's all solvable.

Reply:
Yeah, I originally planned for the API to be stateless, because I assumed that the transaction would take a lock and didn't want an extension crashing or something similar blocking something else from running.

I'm honestly on the fence and wanted to experiment with both options.

What I like about stateful RPCs
* You can do more complex things that might not neccesarily map to a single jj command
* You can perform atomic operations containing lots of things at once
* Potential performance benefits when doing complex things. For example, if we had to perform a bunch of rebases, and then update a bunch of commit descriptions, we could do it in a single transaction, allowing us to skip creating the intermediate commits between the two.

What I like about a 1:1 mapping of RPC <-> jj command
* Users will generally think in terms of jj commands, so it's much easier to use
* Users don't have to implement complex commands like `jj parallelize` themselves

I think it's mostly a question of power vs simplicity. Low-level operations as commands give you a lot of power, but are far more difficult to use. I think, to be honest, looking at it now, we need those high-level operations (if a user had to implement parallelize themselves, they're just gonna use the jj cli instead of the jj api).

We could potentially implement both sets, with the high-level commands written in terms of those low-level commands, and I think that's probably the best option, but that will introduce a maintainence burden (not sure how much).

Reply:
> I'm honestly on the fence and wanted to experiment with both options.
> 
> What I like about stateful RPCs
> 
> * You can do more complex things that might not neccesarily map to a single jj command
> * You can perform atomic operations containing lots of things at once
> * Potential performance benefits when doing complex things. For example, if we had to perform a bunch of rebases, and then update a bunch of commit descriptions, we could do it in a single transaction, allowing us to skip creating the intermediate commits between the two.
> 
> What I like about a 1:1 mapping of RPC <-> jj command
> 
> * Users will generally think in terms of jj commands, so it's much easier to use
> * Users don't have to implement complex commands like `jj parallelize` themselves
> 
> I think it's mostly a question of power vs simplicity. Low-level operations as commands give you a lot of power, but are far more difficult to use. I think, to be honest, looking at it now, we need those high-level operations (if a user had to implement parallelize themselves, they're just gonna use the jj cli instead of the jj api).

If we take the approach of both providing a low-level API and a CLI like API, they should be separate services with different guarantees to not have such a large API surface on a single service. 

> We could potentially implement both sets, with the high-level commands written in terms of those low-level commands, and I think that's probably the best option, but that will introduce a maintainence burden (not sure how much).

I agree but think that you still should put your sketch of the high level API up, even if it gets superseded by the low-level API for this PR.

Reply:
I agree that we should focus on the high-level API for now. The level I'm thinking of would be roughly matching the CLI. The purpose is to make it easier to implement UIs and IDE extensions.

We can add lower-level functionality later IMO.

Reply:
I don't really care about the design as long as it's usable and I have no clue of any internals here
but as a developer for other projects I can say that what I want from an api *right now* is just something that outputs information from e.g. `jj diff`, `jj st`, `jj log` and allows to use basic commands
this is required to be able to write integrations which are more or less future proof (language independent once that gets implemented) and you don't need to parse ugly text